File: laravel/app/Http/Controllers/Api/Marketplace/OrderController.php
<?php

namespace App\Http\Controllers\Api\Marketplace;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Http\JsonResponse;
use App\Models\Marketplace\Order;
use App\Models\Players\Player;
use App\Models\Marketplace\Coupon;

class OrderController extends Controller
{

    public function index(): JsonResponse
    {
        return response()->json(Order::all());
    }

    public function store(Request $request): JsonResponse
    {
        $validated = $request->validate([
            'status' => 'required|string|in:Pending,Paid,Processing,Shipped,Completed,Cancelled,Refunded|max:20',
            'total' => 'required',
            'discount_applied' => 'required',
            'currency' => 'required|string|max:3|regex:/[A-Z]{3}/',
            'payment_method' => 'nullable|string|in:Card,PayPal,Crypto,PlatformCredits|max:20',
            'payment_reference' => 'nullable|string|max:200',
            'shipping_address' => 'nullable|string|max:200',
            'tracking_number' => 'required_if:status,Shipped|nullable|string|max:100',
            'created_at' => 'required|date',
            'paid_at' => 'required_if:status,Paid|nullable|date',
            'shipped_at' => 'nullable|date',
            'player_id' => 'required|exists:players,id',
            'coupon_id' => 'nullable|exists:coupons,id',
        ]);
        $item = Order::create($validated);
        $item->validateRules();
        try {
            $item->validateImplies();
        } catch (\RuntimeException $e) {
            return response()->json(['error' => $e->getMessage()], 422);
        }

        return response()->json($item, 201);
    }

    public function show(Order $order): JsonResponse
    {
        if ($order->player_id !== auth()->id()) {
            abort(403, 'You do not own this resource.');
        }
        return response()->json($order);
    }

    public function cancel(Request $request, Order $order): JsonResponse
    {
        $order->cancel();
        $order->save();
        return response()->json(null, 204);
    }

    public function pay(Request $request, Order $order): JsonResponse
    {
        if (!($order->status === 'Pending')) {
            return response()->json(['error' => 'Guard condition not met for pay'], 422);
        }
        $payment_ref = $request->input('payment_ref');
        $result = $order->pay($payment_ref);
        $order->save();
        return response()->json(['result' => $result]);
    }

    public function processPayment(Request $request, Order $order): JsonResponse
    {
        $result = $order->processPayment();
        $order->save();
        return response()->json(['result' => $result]);
    }

    public function calculateTotal(Request $request, Order $order): JsonResponse
    {
        $result = $order->calculateTotal();
        $order->save();
        return response()->json(['result' => $result]);
    }

    public function applyDiscount(Request $request, Order $order): JsonResponse
    {
        $percent = $request->input('percent');
        $result = $order->applyDiscount($percent);
        $order->save();
        return response()->json(['result' => $result]);
    }

    public function refund(Request $request, Order $order): JsonResponse
    {
        $order->refund();
        $order->save();
        return response()->json(null, 204);
    }
    public function transitionPendingToPaid(Order $order): JsonResponse
    {
        try {
            $order->assertTransition('Paid');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        try {
            if ($order->payment_method === null) {
                throw new \RuntimeException('payment_method is required for Pending -> Paid');
            }
        } catch (\RuntimeException $e) {
            return response()->json(['error' => $e->getMessage()], 422);
        }
        $order->status = 'Paid';
        $order->processPayment(); // @after
        $order->save();
        return response()->json($order);
    }

    public function transitionPaidToProcessing(Order $order): JsonResponse
    {
        if (!in_array(auth()->user()?->role, ['Admin', 'Staff'], true)) {
            return response()->json(['error' => 'Insufficient role for transition Paid -> Processing'], 403);
        }
        try {
            $order->assertTransition('Processing');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        $order->status = 'Processing';
        $order->save();
        return response()->json($order);
    }

    public function transitionProcessingToShipped(Order $order): JsonResponse
    {
        if (!in_array(auth()->user()?->role, ['Admin', 'Staff'], true)) {
            return response()->json(['error' => 'Insufficient role for transition Processing -> Shipped'], 403);
        }
        try {
            $order->assertTransition('Shipped');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        try {
            if ($order->tracking_number === null) {
                throw new \RuntimeException('tracking_number is required for Processing -> Shipped');
            }
        } catch (\RuntimeException $e) {
            return response()->json(['error' => $e->getMessage()], 422);
        }
        $order->status = 'Shipped';
        $order->notifyShipped(); // @after
        $order->save();
        return response()->json($order);
    }

    public function transitionShippedToCompleted(Order $order): JsonResponse
    {
        if (!in_array(auth()->user()?->role, ['Admin', 'Staff'], true)) {
            return response()->json(['error' => 'Insufficient role for transition Shipped -> Completed'], 403);
        }
        try {
            $order->assertTransition('Completed');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        $order->status = 'Completed';
        $order->save();
        return response()->json($order);
    }

    public function transitionPendingToCancelled(Order $order): JsonResponse
    {
        try {
            $order->assertTransition('Cancelled');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        $order->status = 'Cancelled';
        $order->cancel(); // @after
        $order->save();
        return response()->json($order);
    }

    public function transitionPaidToCancelled(Order $order): JsonResponse
    {
        if (!in_array(auth()->user()?->role, ['Admin', 'Staff'], true)) {
            return response()->json(['error' => 'Insufficient role for transition Paid -> Cancelled'], 403);
        }
        try {
            $order->assertTransition('Cancelled');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        $order->status = 'Cancelled';
        $order->cancel(); // @after
        $order->save();
        return response()->json($order);
    }

    public function transitionCompletedToRefunded(Order $order): JsonResponse
    {
        if (!in_array(auth()->user()?->role, ['Admin'], true)) {
            return response()->json(['error' => 'Insufficient role for transition Completed -> Refunded'], 403);
        }
        try {
            $order->assertTransition('Refunded');
        } catch (\InvalidArgumentException $e) {
            return response()->json(['error' => $e->getMessage()], 409);
        }
        $order->status = 'Refunded';
        $order->refund(); // @after
        $order->save();
        return response()->json($order);
    }

    public function transitionRefundedToCompleted(Order $order): JsonResponse
    {
        return response()->json(['error' => 'Transition Refunded -> Completed is not allowed'], 409);
    }

    public function transitionCompletedToCancelled(Order $order): JsonResponse
    {
        return response()->json(['error' => 'Transition Completed -> Cancelled is not allowed'], 409);
    }
}

File: phoenix/lib/cards_project_web/controllers/marketplace/order_controller.ex
defmodule CardsProjectWeb.Marketplace.OrderController do
  use Phoenix.Controller, formats: [:json]
  import Plug.Conn

  alias CardsProject.Marketplace
  alias CardsProject.Marketplace.Order

  def index(conn, _params) do
    orders = Marketplace.list_orders()
    json(conn, Enum.map(orders, &serialize_order/1))
  end

  def show(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    json(conn, serialize_order(order))
  end

  def create(conn, params) do
    case Marketplace.create_order(params) do
      {:ok, order} ->
        conn
        |> put_status(:created)
        |> json(serialize_order(order))

      {:error, changeset} ->
        conn
        |> put_status(:unprocessable_entity)
        |> json(%{errors: format_errors(changeset)})
    end
  end

  # DELETE /api/orders/{id}/cancel
  def cancel(conn, %{"id" => id}) do
    Marketplace.order_cancel_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # POST /api/orders/{id}/pay
  def pay(conn, %{"id" => id} = params) do
    payment_ref = Map.get(params, "payment_ref")
    result = Marketplace.order_pay_behavior(id, payment_ref)
    json(conn, %{result: result})
  end

  # POST /api/orders/{id}/process-payment
  def process_payment(conn, %{"id" => id}) do
    result = Marketplace.order_process_payment_behavior(id)
    json(conn, %{result: result})
  end

  # GET /api/orders/{id}/total
  def calculate_total(conn, %{"id" => id}) do
    result = Marketplace.order_calculate_total_behavior(id)
    json(conn, %{result: result})
  end

  # PATCH /api/orders/{id}/discount
  def apply_discount(conn, %{"id" => id} = params) do
    percent = Map.get(params, "percent")
    result = Marketplace.order_apply_discount_behavior(id, percent)
    json(conn, %{result: result})
  end

  # POST /api/orders/{id}/refund
  def refund(conn, %{"id" => id}) do
    Marketplace.order_refund_behavior(id)
    send_resp(conn, :no_content, "")
  end

  # PATCH /api/orders/:id/transitions/pending-to-paid
  def transition_pending_to_paid(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_pending_to_paid_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/paid-to-processing
  def transition_paid_to_processing(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_paid_to_processing_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/processing-to-shipped
  def transition_processing_to_shipped(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_processing_to_shipped_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/shipped-to-completed
  def transition_shipped_to_completed(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_shipped_to_completed_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/pending-to-cancelled
  def transition_pending_to_cancelled(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_pending_to_cancelled_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/paid-to-cancelled
  def transition_paid_to_cancelled(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_paid_to_cancelled_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/completed-to-refunded
  def transition_completed_to_refunded(conn, %{"id" => id}) do
    order = Marketplace.get_order!(id)
    case Marketplace.transition_completed_to_refunded_order(order) do
      {:ok, updated} ->
        json(conn, serialize_order(updated))

      {:error, :conflict, msg} ->
        conn |> put_status(:conflict) |> json(%{error: msg})

      {:error, :unprocessable, msg} ->
        conn |> put_status(:unprocessable_entity) |> json(%{error: msg})

      {:error, changeset} ->
        conn |> put_status(:unprocessable_entity) |> json(%{errors: format_errors(changeset)})
    end
  end

  # PATCH /api/orders/:id/transitions/refunded-to-completed
  def transition_refunded_to_completed(conn, %{"id" => _id}) do
    conn
    |> put_status(:conflict)
    |> json(%{error: "Transition Refunded -> Completed is not allowed"})
  end

  # PATCH /api/orders/:id/transitions/completed-to-cancelled
  def transition_completed_to_cancelled(conn, %{"id" => _id}) do
    conn
    |> put_status(:conflict)
    |> json(%{error: "Transition Completed -> Cancelled is not allowed"})
  end

  defp serialize_order(%Order{} = record) do
    record
    |> Map.take([:id, :status, :total, :discount_applied, :currency, :payment_method, :payment_reference, :shipping_address, :tracking_number, :created_at, :paid_at, :shipped_at, :player_id, :items_id, :coupon_id])
    |> (fn m -> Map.put(Map.delete(m, :created_at), :created_at, Map.get(m, :created_at)) end).()
    |> (fn m -> Map.put(Map.delete(m, :paid_at), :paid_at, Map.get(m, :paid_at)) end).()
    |> (fn m -> Map.put(Map.delete(m, :shipped_at), :shipped_at, Map.get(m, :shipped_at)) end).()
  end

  defp format_errors(changeset) do
    Ecto.Changeset.traverse_errors(changeset, fn {msg, opts} ->
      Regex.replace(~r"%{(\w+)}", msg, fn _, key ->
        opts |> Keyword.get(String.to_existing_atom(key), key) |> to_string()
      end)
    end)
  end
end
